packages/bff/src/services/purposeService.ts

```diff
@@ -4,16 +4,277 @@ import {
+    tenantClient.tenant.getTenant({
+      params: {
+        id,
+      },
+      withCredentials: true,
+      headers,
+    });
+  const consumers = await Promise.all(
+    [...new Set(purposes.results.map((p) => p.consumerId))].map(getTenant)
+  );
+  const producers = await Promise.all(
+    [...new Set(eservices.map((e) => e.producerId))].map(getTenant)
+  );
+
+  const enhancePurpose = async (
+    purpose: purposeApi.Purpose
+  ): Promise<bffApi.Purpose> => {
+    const eservice = eservices.find((e) => e.id === purpose.eserviceId);
+    if (!eservice) {
+      throw eServiceNotFound(unsafeBrandId(purpose.eserviceId));
+    }
+
+    const producer = producers.find((p) => p.id === eservice.producerId);
+    if (!producer) {
+      throw tenantNotFound(unsafeBrandId(eservice.producerId));
+    }
+
+    const consumer = consumers.find((c) => c.id === purpose.consumerId);
+    if (!consumer) {
+      throw tenantNotFound(unsafeBrandId(purpose.consumerId));
+    }
+
+    const latestAgreement = await getLatestAgreement(
+      agreementClient,
+      purpose.consumerId,
+      eservice,
+      headers
+    );
+    if (!latestAgreement) {
+      throw agreementNotFound(unsafeBrandId(purpose.consumerId));
+    }
+
+    const currentDescriptor = eservice.descriptors.find(
+      (d) => d.id === latestAgreement.descriptorId
+    );
+    if (!currentDescriptor) {
+      throw eServiceDescriptorNotFound(
+        unsafeBrandId(eservice.id),
+        unsafeBrandId(latestAgreement.descriptorId)
+      );
+    }
+
+    const clients =
+      requesterId === purpose.consumerId
+        ? (
+            await getAllClients(
+              authorizationClient,
+              purpose.consumerId,
+              purpose.id,
+              headers
+            )
+          ).map((c) => ({
+            id: c.client.id,
+            name: c.client.name,
+            hasKeys: c.keys.length > 0,
+          }))
+        : [];
+
+    const currentVersion = getCurrentVersion(purpose.versions);
+    const waitingForApprovalVersion = purpose.versions.find(
+      (v) => v.state === "WAITING_FOR_APPROVAL"
+    );
+    // eslint-disable-next-line functional/immutable-data
+    const latestVersion = purpose.versions
+      .sort(
+        (a, b) =>
+          new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
+      )
+      .pop();
+
+    const rejectedVersion =
+      latestVersion?.state === "REJECTED" ? latestVersion : undefined;
+
+    const isUpgradable = (
+      descriptor: catalogApi.EServiceDescriptor,
+      agreement: agreementApi.Agreement,
+      descriptors: catalogApi.EServiceDescriptor[]
+    ): boolean =>
+      descriptors
+        .filter((d) => Number(d.version) > Number(descriptor.version))
+        .some(
+          (d) =>
+            (d.state === "PUBLISHED" || d.state === "SUSPENDED") &&
+            (agreement.state === "ACTIVE" || agreement.state === "SUSPENDED")
+        );
+
+    return {
```

Comment:
Could you extract this in a converter in `purposeApiConverter.ts`? It will also be used in get purpose I think and maybe in other parts of the BFF